### PR TITLE
[SPARK-47094][SQL] SPJ : Dynamically rebalance number of buckets when they are not equal

### PR DESCRIPTION
**Pull request by szehon-ho**
### What changes were proposed in this pull request?
-- Allow SPJ between 'compatible' bucket funtions
-- Add a mechanism to define 'reducible' functions, one function whose output can be 'reduced' to another for all inputs.

  ### Why are the changes needed?
-- SPJ currently applies only if the partition transform expressions on both sides are identifical.

  ### Does this PR introduce _any_ user-facing change?
No

  ### How was this patch tested?
Added new tests in KeyGroupedPartitioningSuite

  ### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -1537,6 +1537,18 @@ object SQLConf {
       .booleanConf
       .createWithDefault(false)
 
+  val V2_BUCKETING_ALLOW_COMPATIBLE_TRANSFORMS =
+    buildConf("spark.sql.sources.v2.bucketing.allow.enabled")
```
Comment (himadripal) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1541]: may be rename `spark.sql.sources.v2.bucketing.allow.enabled` to `spark.sql.sources.v2.bucketing.allowCompatibleTransform.enabled`

Comment (szehon-ho) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Thanks, yea it is a copy and paste error.

**Comment by himadripal** [comment]
was trying this test case, 
`val partition1 = Array(Expressions.years("ts"),
  bucket(2, "id"))
val partition2 = Array(Expressions.days("ts"),
  bucket(4, "id"))`

this throws an exception when paritiallyClustered = false and allowCompatibleTransform=true. 

exception :
`class java.lang.Integer cannot be cast to class java.lang.Long (java.lang.Integer and java.lang.Long are in module java.base of loader 'bootstrap')
java.lang.ClassCastException: class java.lang.Integer cannot be cast to class java.lang.Long (java.lang.Integer and java.lang.Long are in module java.base of loader 'bootstrap')
	at scala.runtime.BoxesRunTime.unboxToLong(BoxesRunTime.java:103)
	at org.apache.spark.sql.catalyst.expressions.BaseGenericInternalRow.getLong(rows.scala:42)`

**Comment by sunchao** [comment]
Thanks @szehon-ho ! will take a look in a few days.

**Comment by szehon-ho** [comment]
Thanks @himadripal I was able to modify one of my tests to also reproduce it, and it should be fixed now.

**Review by advancedxy — COMMENTED** [review]
Hi @szehon-ho,  great to see this feature landing on the Spark side.

To be honest, I am implementing a similar feature called bucket rescaling, but on the Iceberg side. By bucket rescaling, it means that bucketed Iceberg table could report its partitioning with an on-demand bucket number(compatible with current spec's bucket number, a.k.a `on_demand_bucket_number % current_bucket_num = 0 or current_bucket_number % on_demand_bucket_number = 0`). It's also possible for the bucketed iceberg table to have mixed but compatible bucket numbers in its specs.

So, I'd like to share two cents about this topic, hopes that it can share some insights. 
1. besides down scale a bucket, it's also possible to upscale a bucket as long as we put a `bucket_id filter` after the the scan node. For example, to upscale bucket number from 4 to 8 for the bucket_id = 3 bucket. It could be divided into two buckets: `bucket_id_8=3` with filter `bucket(8, id) = 3` and `bucket_id_8 = 7` with filter `bucket(8, id) = 7`
2. it's possible to support bucket number evolution in one bucket partition field.

Anyway, it's game change for Spark's SPJ to support compatible bucket numbers.  The up scale cases might be supported in follow-up prs.

File: sql/core/src/test/scala/org/apache/spark/sql/connector/KeyGroupedPartitioningSuite.scala [hunk 2]
```diff
@@ -1310,6 +1314,312 @@ class KeyGroupedPartitioningSuite extends DistributionAndOrderingSuiteBase {
...
+               |ORDER BY t1.store_id, t1.dept_id, t1.data, t2.data
+               |""".stripMargin)
+
+          val shuffles = collectShuffles(df.queryExecution.executedPlan)
+          assert(shuffles.isEmpty, "SPJ should be triggered")
+
+          val scans = collectScans(df.queryExecution.executedPlan).map(_.inputRDD.
+            partitions.length)
+
+          val expectedBuckets = Math.min(table1buckets, table2buckets)
```
[the remark this answers is not in the document]
Comment (szehon-ho) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/KeyGroupedPartitioningSuite.scala:1540]: Yes, in fact it is definitely a follow up to do this with 'partiallyClustered' (which is currently not enabled).  This mode today 'duplicate' partitions for the partially clustered side, and I think can be used to turn on the same for compatible transforms.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala [hunk 3]
```diff
@@ -635,6 +636,22 @@ trait ShuffleSpec {
...
+   * 'reducible' on the corresponding partition expression function of the other shuffle spec.
+   * <p>
+   * If a value is returned, there must be one Option[[Reducer]] per partition expression.
+   * A None value in the set indicates that the particular partition expression is not reducible
+   * on the corresponding expression on the other shuffle spec.
+   * <p>
+   * Returning none also indicates that none of the partition expressions can be reduced on the
+   * corresponding expression on the other shuffle spec.
+   */
+  def reducers(spec: ShuffleSpec): Option[Seq[Option[Reducer[_]]]] = None
```
Comment (sunchao) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:654]: maybe this should belong to `KeyGroupedShuffleSpec`? I'm not sure whether this API is meaningful for other shuffle specs.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala [hunk 4]
```diff
@@ -829,20 +846,59 @@ case class KeyGroupedShuffleSpec(
     }
   }
 
+  override def canCreatePartitioning: Boolean = SQLConf.get.v2BucketingShuffleEnabled &&
```
Comment (sunchao) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:849]: nit: seems unnecessary change

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/functions/ReducibleFunction.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+/**
+ * Base class for user-defined functions that can be 'reduced' on another function.
+ *
+ * A function f_source(x) is 'reducible' on another function f_target(x) if
+ * there exists a reducer function r(x) such that r(f_source(x)) = f_target(x) for all input x.
+ *
+ * @since 4.0.0
+ */
+@Evolving
+public interface ReducibleFunction<T, A> extends ScalarFunction<T> {
```
Comment (sunchao) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/functions/ReducibleFunction.java:31]: what is `T` and what is `A`? should `Reducer` also have two type parameters?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala [hunk 6]
```diff
@@ -829,20 +846,59 @@ case class KeyGroupedShuffleSpec(
...
+
+  override def createPartitioning(clustering: Seq[Expression]): Partitioning = {
+    KeyGroupedPartitioning(clustering, partitioning.numPartitions, partitioning.partitionValues)
+  }
+
+  override def reducers(other: ShuffleSpec): Option[Seq[Option[Reducer[_]]]] = {
+    other match {
+      case otherSpec: KeyGroupedShuffleSpec =>
+        val results = partitioning.expressions.zip(otherSpec.partitioning.expressions).map {
+          case (e1: TransformExpression, e2: TransformExpression)
```
Comment (sunchao) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:861]: maybe it's better to move this logic into `TransformExpression` itself, e.g., add a `TransformExpression.reducer` method.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (sunchao) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:654]: One use case I think could be `hours(col)` vs `days(col)`, in this case we can "reduce" the former into the latter.

File: sql/core/src/test/scala/org/apache/spark/sql/connector/KeyGroupedPartitioningSuite.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (sunchao) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/KeyGroupedPartitioningSuite.scala:1540]: Yes, I think it will be very useful to make partially clustered distribution work with the feature introduced in this PR.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/functions/ReducibleFunction.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+    /**
+     * If this function is 'reducible' on another function, return the {@link Reducer} function.
+     * @param other other function
+     * @param thisArgument argument for this function instance
```
[the remark this answers is not in the document]
Comment (szehon-ho) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/functions/ReducibleFunction.java]: Added javadocs

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala (hunk 3 above)
Comment (szehon-ho) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala]: done

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/functions/ReducibleFunction.java (hunk 5 above)
Comment (szehon-ho) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/functions/ReducibleFunction.java]: added javadocs

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala (hunk 6 above)
Comment (szehon-ho) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala]: done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala (hunk 4 above)
Comment (szehon-ho) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala]: done, i think i was trying to move the private method to the bottom

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/functions/ReducibleFunction.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+     *     <li>other function = bucket(2, x)
+     * </ul>
+     * Invoke with arguments
+     * <ul>
+     *     <li>other = bucket</li>
+     *     <li>this param = Int(4)</li>
+     *     <li>other param = Int(2)</li>
+     * </ul>
+     * @param other the other function
+     * @param thisParam param for this function
```
[the remark this answers is not in the document]
Comment (szehon-ho) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/functions/ReducibleFunction.java]: OK, made default impl for both so that implementations can pick one to override

[the remark this answers is not in the document]
Comment (szehon-ho) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/functions/ReducibleFunction.java]: Actually on second thought, we do plan to add geo partition transforms , for example xz2 ordering https://github.com/spatialx-project/geolake/blob/main/api/src/main/java/org/apache/iceberg/transforms/ExtendedZCurve.java which need resolution.

as well as multi-arg bucket transforms: https://docs.google.com/document/d/1aDoZqRgvDOOUVAGhvKZbp5vFstjsAMY4EFCyjlxpaaw/ 

So i added back the type to Object in latest pr, hope it is ok. 

Note, I tried awhile to make the method type-parameterized like

```
<T> Option<Reducer<I, O>> reducer(ReducibleFunction<I, O> other, T numBuckets,
                                  T otherNumBuckets);
```

 but was not able to override it successfully in Bucket in scala

**Comment by szehon-ho** [comment]
@sunchao can you take another look?  I guess open question is how to support other potential parameterized transforms in the future, let me know your thoughts.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala [hunk 9]
```diff
@@ -846,6 +879,20 @@ case class KeyGroupedShuffleSpec(
   }
 }
 
+object KeyGroupedShuffleSpec {
+  def reducePartitionValue(row: InternalRow,
```
Comment (sunchao) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:883]: nit: we usually follow this style:

```scala
  def reducePartitionValue(
      row: InternalRow,
      expressions: Seq[Expression],
      reducers: Seq[Option[Reducer[_, _]]]):
    InternalRowComparableWrapper = {
    val partitionVals = row.toSeq(expressions.map(_.dataType))
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 10]
```diff
@@ -1541,6 +1541,18 @@ object SQLConf {
       .booleanConf
       .createWithDefault(false)
 
+  val V2_BUCKETING_ALLOW_COMPATIBLE_TRANSFORMS =
+    buildConf("spark.sql.sources.v2.bucketing.allowCompatibleTransforms.enabled")
+      .doc("Whether to allow storage-partition join in the case where the partition transforms" +
```
Comment (sunchao) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1546]: nit: space after `transforms`

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/functions/ReducibleFunction.java [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+   * </ul>
+   *
+   * @param thisParam parameter for this function
+   * @param otherFunction the other parameterized function
+   * @param otherParam parameter for the other function
+   * @return a reduction function if it is reducible, null if not
+   */
+  default Reducer<I, O> reducer(
+    Object thisParam,
+    ReducibleFunction<?, ?> otherFunction,
```
Comment (sunchao) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/functions/ReducibleFunction.java:84]: `ReducibleFunction<I, O> otherFunction` did not work? just wonder why

**Review by sunchao — COMMENTED** [review]
Thanks @szehon-ho , looks much better now. I mainly have one more question on the params.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/functions/ReducibleFunction.java (hunk 11 above)
Comment (szehon-ho) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/functions/ReducibleFunction.java:84]: Yea I think I get an error like:

```
/Users/szehon/repos/apache-spark/spark/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TransformExpression.scala:105:42: type mismatch;
[error]  found   : org.apache.spark.sql.connector.catalog.functions.ReducibleFunction[_$5,_$6] where type _$6, type _$5
[error]  required: org.apache.spark.sql.connector.catalog.functions.ReducibleFunction[_$3,_$4]
[error]         thisFunction.reducer(numBuckets, otherFunction, otherNumBuckets)
[error]                                          ^
[error] /Users/szehon/repos/apache-spark/spark/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TransformExpression.scala:106:38: type mismatch;
[error]  found   : org.apache.spark.sql.connector.catalog.functions.ReducibleFunction[_$5,_$6] where type _$6, type _$5
```

But actually there is no requirement they have to be the exact same Input, Output type, in theory the other function could have different types.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/functions/ReducibleFunction.java [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+   * <p>
+   * Example to return reducer for reducing f_source = bucket(4, x) on f_target = bucket(2, x)
+   * <ul>
+   *     <li>thisFunction = bucket</li>
+   *     <li>thisParam = Int(4)</li>
+   *     <li>otherFunction = bucket</li>
+   *     <li>otherParam = Int(2)</li>
+   * </ul>
+   *
+   * @param thisParam parameter for this function
```
[the remark this answers is not in the document]
Comment (szehon-ho) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/functions/ReducibleFunction.java:77]: I was thinking a list, would that work or cause any issue ?

**Comment by szehon-ho** [comment]
@sunchao thanks ! addressed review comments.  Lmk if supporting generic function parameter is too messy, and we want to switch back to just bucket for first cut.

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/functions/ReducibleFunction.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (sunchao) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/functions/ReducibleFunction.java:77]: I'm thinking whether we can use `InternalRow` for this, which is consistent with the v2 function API (see `ScalarFunction` and `AggregateFunction`, and can support multiple parameters.

We need to define a clear contract between Spark and the callers of this method though: **what should a caller expect to see from `thisParam` and `otherParam`**. 

For `bucket` it is quite clear, but I'm not sure how this will this work with geohash function that you pointed out. In Spark, we need a way for the data source provider to pass the `resolution` of the function to Spark via v2 catalog, and then pass it back to the data source via `reducer`. Any idea?

[the remark this answers is not in the document]
Comment (szehon-ho) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/functions/ReducibleFunction.java:77]: Yea I think you are right, it requires more changes to Spark, V2ExpressionUtil, etc to add these.  Let me revert back to bucketReducer() and then later if we figure that out, we can make this just call the general case.

**Comment by szehon-ho** [comment]
@sunchao if you have time for another look, reverted to use specific argument type and method for bucket, and worry about other parameterized transforms later.

**Review by sunchao — COMMENTED** [review]
Looks mostly good!

**Comment by szehon-ho** [comment]
@sunchao thanks can you take another look?  I think the failed test is not related (looks like error in pyspark uploading test report).

**Review by sunchao — APPROVED** [review]
LGTM

**Review by himadripal — APPROVED** [review]
Thank you for adding this feature. @szehon-ho

**Comment by sunchao** [comment]
Thanks, merged to master!

**Comment by szehon-ho** [comment]
Thanks @sunchao and all for the review!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala [hunk 13]
```diff
@@ -164,6 +165,18 @@ case class BatchScanExec(
               (groupedParts, expressions)
           }
 
+          // Also re-group the partitions if we are reducing compatible partition expressions
+          val finalGroupedPartitions = spjParams.reducers match {
```
Comment (rekbun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala:169]: I believe this could produce incorrect result when joining presorted bucketed tables with compatible bucket counts. 

Specifically, if we have two tables:
1. Bucketed and sorted on the same join keys
2. With different bucket counts, where one table's bucket count is a multiple of the other


When performing a bucketed join in Spark, it's expected that the sort order should be preserved. However, it appears that the merging or grouping process involved in the join might break these sorting guarantees, leading to incorrect results.
